src/Features/Core/Portable/Snippets/AbstractSnippetService.cs

```diff
@@ -46,10 +49,23 @@ public ISnippetProvider GetSnippetProvider(string snippetIdentifier)
         /// </summary>
         public async Task<ImmutableArray<SnippetData>> GetSnippetsAsync(Document document, int position, CancellationToken cancellationToken)
         {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            if (syntaxFacts.IsInNonUserCode(syntaxTree, position, cancellationToken))
+            {
+                return ImmutableArray<SnippetData>.Empty;
+            }
+
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+            var context = document.GetRequiredLanguageService<ISyntaxContextService>().CreateContext(document, semanticModel, position, cancellationToken);
+
             using var _ = ArrayBuilder<SnippetData>.GetInstance(out var arrayBuilder);
             foreach (var provider in GetSnippetProviders(document))
             {
-                var snippetData = await provider.GetSnippetDataAsync(document, position, cancellationToken).ConfigureAwait(false);
+                var snippetData = provider is AbstractSnippetProvider abstractSnippetProvider
+                    ? abstractSnippetProvider.GetSnippetDataCore(context, cancellationToken)
+                    : await provider.GetSnippetDataAsync(document, position, cancellationToken).ConfigureAwait(false);
```

Comment:
`GetSnippetDataCore` will always be used for C#/VB. Keeping the `GetSnippetDataAsync` and the interface as-is for other languages.

Reply:
but this is AbstractSnippetService, which is only for C#/VB anyways.